chainer/link.py

```diff
@@ -245,30 +245,35 @@ def copy(self):
             d[name].grad = None
         return ret
 
-    def to_cpu(self):
+    def to_cpu(self, copy=False):
         """Copies parameter variables and persistent values to CPU.
 
         This method does not handle non-registered attributes. If some of such
         attributes must be copied to CPU, the link implementation must
         override this method to do so.
 
-        Returns: self
+        Returns: self if `copy`` is False, otherwise a copied link
```

Comment:
Thanks! Compared with `Variable` implementation of Chainer v1 and #2072, this change introducing `copy=True` essentially does the same thing with `link.copy()` followed by `link.to_cpu()` and looks making no sense.

Under the change of `Variable` implementation in this PR, introducing `copy=True` makes it possible to efficiently copy an array from CPU to GPU with leaving a link accessible  to the original array on CPU without extra CPU to CPU copy. This requirement is caused by the change that how `Variable` has an array internally and it would not be quite clear for users, as you say.

In this PR, `Variable` has the following three states and it would cause this problem:
- uninitialized
- initialized, array on CPU
- initialized, array on GPU

I will think another model where `Variable` has only two states:
- uninitialized
- initialized

and Variable that has its array on CPU is another instance from one that has its array on GPU, which may be more natural for Chainer.

Let me implement the second model.